apluslms_roman/cli.py

```diff
@@ -365,20 +416,136 @@ def build_action(context):
+        except ValidationError as e:
+            exit(1, '\n'.join(render_error(e)))
+    print(_("Project configuration file {} created successfully.".format(name)))
+
+
 def config_print_action(context):
     all_ = not any(getattr(context.args, k, False) for k in ('user', 'project'))
     if context.args.debug:
         print("---\n# arguments:")
         data = {k: v for k, v in vars(context.args).items() if k[0] != '_' and not callable(v)}
         yaml_dump(data, stdout)
-    if all_ or context.args.user:
+    if all_ or context.args.global_:
         print("---\n# roman settings:")
         yaml_dump(context.settings._data, stdout)
     if all_ or context.args.project:
         print("---\n# project config")
         yaml_dump(get_config(context)._data, stdout)
 
 
+def config_set_action(context):
+    if context.args.global_ == context.args.project:
+        raise Exception("Choose either global settings or project settings")
+    document = context.settings if context.args.global_ else get_config(context)
+
+    for val in context.args.values:
+        try:
+            key, val = val.split('=', 1)
+            document.mlset_cast(key, val)
+        except ValueError as err:
+            if hasattr(err, 'value_type'):
+                raise ValueError(_("{} should be of type '{}', but was '{}'.").format(
+                    key, err.value_type, type(val).__name__))
+            else:
+                raise ValueError(_("Give values in format 'key=val'."))
+
+    document.validate()
+    document.save()
+
+
+def config_rm_action(context):
+    if context.args.global_ == context.args.project:
+        raise Exception("Choose either global settings or project settings")
+    document = context.settings if context.args.global_ else get_config(context)
+
+    for val in context.args.keys:
+        document.mldel(val)
+
+    document.validate()
+    document.save()
+
+
+def step_list_action(context):
+    steps = get_config(context).steps
+    num_len = max(2, len(str(len(steps)-1)))
+    name_len = max(4, max(len(s.get('name', '')) for s in steps))
+    header_fmt = "{:>%ds}  {:%ds} {}" % (num_len, name_len)
+    step_fmt = "{:%dd}. {:%ds} {}" % (num_len, name_len)
+    print(header_fmt.format('ID', 'NAME', 'IMAGE'))
+    for i, step in enumerate(steps):
+        print(step_fmt.format(i, step.get('name', ''), step['img']))
+    return
+
+
+def step_add_action(context):
+    args = context.args
+    env = args.env
+    if env:
+        try:
+            env = {a: b for a, b in [s.split('=', 1) for s in env]}
+        except ValueError:
+            raise ValueError("env is a dict, so values need to be in key=val format, e.g. a=1 b=2")
+    step = {
+        'img': args.img,
+        'cmd': args.cmd,
+        'mnt': args.mnt,
+        'env': env,
+        'name': args.name
+    }
+    step = {k: v for k, v in step.items() if v}
+    config = get_config(context)
+    config.add_step(step)
```

Comment:
Not sure if `config.add_step` should use definition `def add_step(img, cmd=None, mnt=None, env=None, name=None)` instead. This would require to call it `config.add_step(**step)`, but then the python interface would ensure that only valid values are added to a step. Sadly, then `add_step` needs to filter `None` or false values on that side too.

Any thoughts?

Reply:
I don't really understand what you mean with "then the python interface would ensure that only valid values are added to a step"...

Reply:
`add_steps(**a_dict)` -> `def add_step(img, cmd=None, mnt=None, env=None, name=None)` conversion checks that there are only keys `img`, `cmd`, `mnt`, `env` and `name`. Any other key would raise error.

Alternative would be for `add_step` to validate that the dictionary doesn't contain any invalid keys.

Though, now that I think about it, this might introduce unnesesary duplication of logic as schema ensures the dictionary is valid in the end.

If the GUI benefits from a different type of interface, then someone can add that.

Thus, after rethinking, let's go with what there is now.